### PR TITLE
Fix Duplicate Cart Items in Explorer

### PR DESCRIPTION
**Pull request by boudra**
This fixes the cart allowing duplicate projects.

## PR checklist

For every PR, make sure that these statements are true:
- [x] Includes only changes relevant to the original ticket. Significant refactoring needs to be separated.
- [x] Doesn't contain type casts and non-null assertions.
- [x] Doesn't add `@ts-ignore`.
- [x] Doesn't disable lints.
- [x] Doesn't use `useState` just for computation - use plain variables instead.
- [x] Splits components into pure components that don't depend on external state or hooks.
- [x] Avoid embedding components within other components
- [x] Doesn't propagate optional values without good reason, doesn't mark property values as optional if that doesn't represent reality.
- [x] Doesn't duplicate existing code.
- [x] Parses out-of-domain data - this includes user input, API respones, on-chain data etc.
- [x] Doesn't contain commented out code.
- [x] Doesn't contain skipped or empty tests.
- [x] If this PR adds/updates any feature, it adds/updates its test script

Subjective - at the discretion of the reviewers
- Does things as simply as possible, but not simpler.
- Doesn't reinvent the wheel or create premature abstractions.

##### Refers/Fixes

fixes #issuenumber

### CODE REVIEW

File: packages/grant-explorer/src/store.ts [hunk 1]
```diff
@@ -46,19 +46,49 @@ const defaultVotingTokens = Object.fromEntries(
   })
 ) as Record<ChainId, VotingToken>;
 
+function areProjectIdsEqual(a: CartProject, b: CartProject): boolean {
```
Comment (bard) [inline packages/grant-explorer/src/store.ts:49]: `isSameProject`?

File: packages/grant-explorer/src/store.ts [hunk 2]
```diff
@@ -46,19 +46,49 @@ const defaultVotingTokens = Object.fromEntries(
...
+      };
+    } else {
+      return { ...acc, projects: [...acc.projects, project] };
+    }
+  }, initialAcc);
+
+  return result.hasUpdatedProject
+    ? result.projects
+    : [...currentProjects, newProject];
+}
```
Comment (bard) [inline packages/grant-explorer/src/store.ts:81]: I'm not sure I interpret the above correctly. Could it be done like this?

```ts
function updateOrInsertCartProject(
  currentProjects: CartProject[],
  newProject: CartProject
): CartProject[] {
  if (currentProjects.find(p => p.id === newProject.id) !== undefined) {
    return currentProjects;
  } else {
    return [...currentProjects, newProject];
  }
}
```

**Review by bard — APPROVED** [review]

**Comment by carlbarrdahl** [comment]
A better data model for the cart might be a Map with the applicationRef as key and the amount as value.

**Review by melissa-neira — APPROVED** [review]

File: packages/grant-explorer/src/store.ts (hunk 2 above)
Comment (boudra) [reply packages/grant-explorer/src/store.ts:81]: this snippet doesn't update the project when it already exists, it only appends

**Comment by boudra** [comment]
> A better data model for the cart might be a Map with the applicationRef as key and the amount as value.

for sure, it can be simplified a lot

File: packages/grant-explorer/src/store.ts (hunk 2 above)
Comment (bard) [reply packages/grant-explorer/src/store.ts:81]: I thought a project is a data object? Does it have anything mutable?

**Comment by carlbarrdahl** [comment]
> > A better data model for the cart might be a Map with the applicationRef as key and the amount as value.
> 
> @carlbarrdahl for sure, it can be simplified a lot, not sure if we can implement it in this PR though!

True. 
Could be an opportunity to re-think the cart and explore new ways of allocating funds to projects and do it as a larger 2-week cycle (https://basecamp.com/shapeup)

**Pull request merged**